### PR TITLE
SQL: Enhance error message on filtering check against aggs

### LINKED ISSUE #57125: SQL: use of agg function in HAVING with no grouping emits wrong error message
A(n incorrect) query like `SELECT salary FROM employees HAVING salary = count(1)` will generate an error message, but a misleading one: 
`"Found 1 problem\nline 2:46: Cannot use WHERE filtering on aggregate function [count(1)], use HAVING instead"`.

### PR DESCRIPTION
**Pull request by bpintea**
This PR introduces a distinction between the failures caused by WHERE-filtering with aggs
vs HAVING-filtering with missing aggs and improving the error messaging around that.

Closes #57125.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java [hunk 1]
```diff
@@ -662,11 +662,21 @@ private static void checkGroupingFunctionTarget(GroupingFunction f, Set<Failure>
...
+                        if (filterChild instanceof Project) {
+                            filter.condition().forEachDown(FieldAttribute.class,
+                                fa -> localFailures.add(
+                                    fail(e, "[{}] field must appear in the GROUP BY clause or be used in an aggregate function",
+                                        Expressions.name(fa)))
+                            );
+                        } else {
+                            localFailures.add(
+                                fail(e, "Cannot use WHERE filtering on aggregate function [{}], use HAVING instead", Expressions.name(e)));
+
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:678]: nit: extra new line.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 2]
```diff
@@ -912,7 +912,17 @@ public void testIifWithDifferentResultAndDefaultValueDataTypes() {
...
+            error("SELECT MAX(int) FROM test WHERE MAX(int) > 10 GROUP BY bool"));
+    }
+
+    public void testHavingInAggs() {
+        assertEquals("1:29: [int] field must appear in the GROUP BY clause or be used in an aggregate function",
+            error("SELECT int FROM test HAVING MAX(int) = 0"));
+
+        assertEquals("1:35: [int] field must appear in the GROUP BY clause or be used in an aggregate function",
+                error("SELECT int FROM test HAVING int = count(1)"));
+        // Note: "SELECT int FROM test HAVING int = 1" works, though it normally shouldn't; to correct this out, we'd need to qualify the
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:924]: Maybe add a test with `accept(<query>)` and a TODO? to have it more clear?

**Review by matriv — APPROVED** [review]
LGTM. Left a comment.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 3]
```diff
@@ -912,7 +912,17 @@ public void testIifWithDifferentResultAndDefaultValueDataTypes() {
 
     public void testAggsInWhere() {
         assertEquals("1:33: Cannot use WHERE filtering on aggregate function [MAX(int)], use HAVING instead",
-                error("SELECT MAX(int) FROM test WHERE MAX(int) > 10 GROUP BY bool"));
+            error("SELECT MAX(int) FROM test WHERE MAX(int) > 10 GROUP BY bool"));
+    }
+
+    public void testHavingInAggs() {
+        assertEquals("1:29: [int] field must appear in the GROUP BY clause or be used in an aggregate function",
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:919]: Good catch for queries with implicit grouping.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java [hunk 4]
```diff
@@ -662,11 +662,21 @@ private static void checkGroupingFunctionTarget(GroupingFunction f, Set<Failure>
...
-            if ((filter.child() instanceof Aggregate) == false) {
+            LogicalPlan filterChild = filter.child();
+            if (filterChild instanceof Aggregate == false) {
                 filter.condition().forEachDown(Expression.class, e -> {
                     if (Functions.isAggregate(attributeRefs.getOrDefault(e, e))) {
-                        localFailures.add(
-                            fail(e, "Cannot use WHERE filtering on aggregate function [{}], use HAVING instead", Expressions.name(e)));
+                        if (filterChild instanceof Project) {
+                            filter.condition().forEachDown(FieldAttribute.class,
+                                fa -> localFailures.add(
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:671]: Slight adjustment " or in an aggregate function" (drop `be used`)

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:924]: +1. I'd like to see a few more tests with it working and not working especially when dealing with non-naked expressions such as field inside a scalar function.

**Review by costin — COMMENTED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java (hunk 4 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java]: Adjusted.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: A few more queries added under a new test.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 5]
```diff
@@ -916,13 +916,20 @@ public void testAggsInWhere() {
...
+        assertEquals("1:35: [int] field must appear in the GROUP BY clause or in an aggregate function",
                 error("SELECT int FROM test HAVING int = count(1)"));
-        // Note: "SELECT int FROM test HAVING int = 1" works, though it normally shouldn't; to correct this out, we'd need to qualify the
-        // Filter (WHERE vs HAVING); but this "extra flexibility" shouldn't be harmful atp.
+    }
+
+    public void testHavingAsWhere() {
+        // TODO: this query works, though it normally shouldn't; a check about it could only be enforced if the Filter would be qualified
+        // (WHERE vs HAVING). Otoh, this "extra flexibility" shouldn't be harmful atp.
+        accept("SELECT int FROM test HAVING int = 1");
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:929]: thx!

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**